packets1/subscribe.go

```diff
@@ -106,6 +106,19 @@ func (p *Subscribe) Unpack(buf []byte) error {
 }
 
 func (p Subscribe) String() string {
-	return fmt.Sprintf("SUBSCRIBE(TopicName=%#v, QOS=%d, TopicID=%d, TopicIDType=%d, MessageID=%d, Dup=%t)",
-		string(p.TopicName), p.QOS, p.TopicID, p.TopicIDType, p.messageID, p.DUP())
+	switch p.TopicIDType {
+	case TIT_STRING:
+		return fmt.Sprintf("SUBSCRIBE(TopicName=%#v, QOS=%d, MessageID=%d, Dup=%t)",
+			p.TopicName, p.QOS, p.messageID, p.DUP())
+	case TIT_PREDEFINED:
+		return fmt.Sprintf("SUBSCRIBE(TopicID(p)=%d, QOS=%d, MessageID=%d, Dup=%t)",
+			p.TopicID, p.QOS, p.messageID, p.DUP())
+	case TIT_SHORT:
+		topicName := pkts.DecodeShortTopic(p.TopicID)
+		return fmt.Sprintf("SUBSCRIBE(TopicName(s)=%#v, QOS=%d, MessageID=%d, Dup=%t)",
+			topicName, p.QOS, p.messageID, p.DUP())
+	default:
+		return fmt.Sprintf("SUBSCRIBE(TopicName=%#v, QOS=%d, TopicID=%d, TopicIDType=%d (INVALID!), MessageID=%d, Dup=%t)",
+			p.TopicName, p.QOS, p.TopicID, p.TopicIDType, p.messageID, p.DUP())
+	}
```

Comment:
Are such invalid packets useful in any way? As far as I’m concerned I would just ignore this case as you do e.g. in `Pack()`. Even better, make sure such packets cannot be (easily) constructed and panic in the `default` branches everywhere you test `TopicIDType` (as a form of assertion). The latter is probably something for a separate PR; I’m also not sure whether this is the “Go way” to address this.

The comment applies also to `Unsubscribe`.

Reply:
This is probably a matter for a broader discussion about where and when the validity of the packet's values is tested - and therefore, what parts of the code hold what assertions/invariants.

In the `packets1` package, you don't know where the values came from. Was the packet read from an (untrusted) network connection? Was it created by some 3rd party code using a constructor? Was it created by a constructor but values were modified then?

I think it's not practical (or maybe even not possible) in Go to (provably) always hold an assertion that the values kept in the packet's struct are always correct. Therefore, I think, the code should be able to cope with any values.

Regarding `panic()`: I would rather keep the number of `panic()`s as low as possible (ideally zero). Because, in my view, every `panic()` is a potential attack vector - it's very hard to track all code branches where some malicious user can pass some invalid value which will in the end result in panic (and therefore successful DoS). Yes, you can use `recover()`s in suitable places to fence the untrusted parts of the code but I think it's better to not use `panic()` in the first place ;)

Reply:
It’s a broader discussion indeed :-)

The thing that bothers me most within the context of this PR is inconsistency. In `computeLength()` and `Pack()` it is assumed the packet is valid. I see no reason to deviate from that approach in `String()`. The only place where the `TopicIDType` value is checked is in `Unpack()`, but there it makes sense because it is the place where the packet is built from untrusted data.

You are right it is not possible to maintain struct invariants in a struct with public fields in Go. So you need to trust to the user to some degree. The only thing I’m asking within this PR is to do that consistently.

As for the panics, I understand your view. I just don’t think there is another way to signal assertion violation in Go. (In some other languages one would use an exception.) In my view, all `switch` statements in any code should always be exhaustive and if the language itself doesn’t guarantee this, there should be a `default` branch with some kind of “should not be reached” assertion. This really helps when the list of values changes over time and I’ve seen real-world, hard-to-debug bugs caused by not doing this. But as I said, this is possibly something for a different PR, for this one I’m most concerned with inconsistency of approaches.

Reply:
> In my view, all switch statements in any code should always be exhaustive
Yes, that's precisely the reason why I made `String()` switch exhaustive.

> The thing that bothers me most within the context of this PR is inconsistency.
It is kind of consistent but there's one bug ;)

Can we agree on this?

1. If you can return `error`, check the value and return `error` when illegal value is observed.
2. If you can't return `error` and illegal value doesn't have any harmful consequences, return "anything" (garbage in garbage out - user's responsibility)
3. If you can't return `error` and illegal value is harmful, panic.

`computeLength()` is 2. `Pack()` is buggy (should return `error`). `Unpack` is 1. `String()` is 2.

Reply:
Me and @mprymek had a call and we agreed to change the `String()` as suggested in the PR. The main reason is that handling the invalid `TopicIDType` value explicitly is useful for debugging in cases where one constructs e.g. an invalid packet by mistake. The approach is also consistent with similar code such as `ReturnCode.String()`.

As for the wider discussion about invariants/errors/panics, we found we have different views and didn’t reach a clear conclusion. We may revisit the topic if we encounter a similar situation again.